File: apps/api/app/api/v1/codex_relations.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user, get_db
from app.schemas.codex_relation import CodexRelationCreate, CodexRelationRead, CodexRelationUpdate
from app.services.crud_codex_relation import (
    create_codex_relation,
    delete_codex_relation,
    get_codex_relation,
    list_codex_relations,
    update_codex_relation,
)
from app.services.crud_project import get_project

router = APIRouter(prefix="/projects/{project_id}/codex-relations", tags=["codex_relations"])


async def _get_project_or_404(project_id: uuid.UUID, db: AsyncSession):
    project = await get_project(db, project_id)
    if project is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")
    return project


@router.post("", response_model=CodexRelationRead, status_code=status.HTTP_201_CREATED)
async def create(
    project_id: uuid.UUID,
    data: CodexRelationCreate,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexRelationRead:
    await _get_project_or_404(project_id, db)
    return await create_codex_relation(db, project_id, data)


@router.get("", response_model=list[CodexRelationRead])
async def list_all(
    project_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> list[CodexRelationRead]:
    await _get_project_or_404(project_id, db)
    return await list_codex_relations(db, project_id)


@router.get("/{relation_id}", response_model=CodexRelationRead)
async def get_one(
    project_id: uuid.UUID,
    relation_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexRelationRead:
    await _get_project_or_404(project_id, db)
    relation = await get_codex_relation(db, project_id, relation_id)
    if relation is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Codex relation not found"
        )
    return relation


@router.patch("/{relation_id}", response_model=CodexRelationRead)
async def update(
    project_id: uuid.UUID,
    relation_id: uuid.UUID,
    data: CodexRelationUpdate,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexRelationRead:
    await _get_project_or_404(project_id, db)
    relation = await get_codex_relation(db, project_id, relation_id)
    if relation is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Codex relation not found"
        )
    return await update_codex_relation(db, relation, data)


@router.delete("/{relation_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(
    project_id: uuid.UUID,
    relation_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> None:
    await _get_project_or_404(project_id, db)
    relation = await get_codex_relation(db, project_id, relation_id)
    if relation is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Codex relation not found"
        )
    await delete_codex_relation(db, relation)

File: apps/web/lib/api/codex-relations.ts
/**
 * Typed endpoint functions for the CodexRelation resource (UX-3a relationship
 * graph). Relations are PROJECT-scoped (`apps/api/app/api/v1/codex_relations.py`,
 * prefix `/projects/{project_id}/codex-relations` under `/api/v1`):
 *   GET    /projects/{pid}/codex-relations            → CodexRelationRead[]
 *   POST   /projects/{pid}/codex-relations            → CodexRelationRead (201)
 *   GET    /projects/{pid}/codex-relations/{id}       → CodexRelationRead
 *   PATCH  /projects/{pid}/codex-relations/{id}       → CodexRelationRead
 *   DELETE /projects/{pid}/codex-relations/{id}       → 204
 *
 * A relation is a directed edge between two polymorphic codex entities; the
 * graph resolves each `{entity_type, entity_id}` to a label via the project's
 * codex entries. Each function validates its response with Zod — backend drift
 * throws rather than silently flowing a wrong shape into the graph.
 */
import { apiFetch } from "./client";
import {
  codexRelationListSchema,
  codexRelationReadSchema,
  codexRelationCreateSchema,
  codexRelationUpdateSchema,
  type CodexRelationCreate,
  type CodexRelationRead,
  type CodexRelationUpdate,
} from "./types";

/** List the relations belonging to a project. */
export async function listCodexRelations(
  projectId: string,
): Promise<CodexRelationRead[]> {
  const data = await apiFetch<unknown>(
    `/projects/${projectId}/codex-relations`,
  );
  return codexRelationListSchema.parse(data);
}

/** Create a relation under a project. The payload is validated first. */
export async function createCodexRelation(
  projectId: string,
  input: CodexRelationCreate,
): Promise<CodexRelationRead> {
  const body = codexRelationCreateSchema.parse(input);
  const data = await apiFetch<unknown>(
    `/projects/${projectId}/codex-relations`,
    { method: "POST", body },
  );
  return codexRelationReadSchema.parse(data);
}

/** Patch a relation's label (relation_type / description). */
export async function updateCodexRelation(
  projectId: string,
  relationId: string,
  patch: CodexRelationUpdate,
): Promise<CodexRelationRead> {
  const body = codexRelationUpdateSchema.parse(patch);
  const data = await apiFetch<unknown>(
    `/projects/${projectId}/codex-relations/${relationId}`,
    { method: "PATCH", body },
  );
  return codexRelationReadSchema.parse(data);
}

/** Delete a relation (the backend answers 204; apiFetch returns null). */
export async function deleteCodexRelation(
  projectId: string,
  relationId: string,
): Promise<void> {
  await apiFetch<unknown>(
    `/projects/${projectId}/codex-relations/${relationId}`,
    { method: "DELETE" },
  );
}
